%
%
%
%\VignetteIndexEntry{An R package for visualization and analysis of Cufflinks high-throughput sequencing data}
%\VignetteKeywords{cummeRbund,visualization,sequencing,cufflinks}
%\VignettePackage{cummeRbund}
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\documentclass[10pt]{article}
\usepackage{amsmath}
\usepackage[authoryear,round]{natbib}
\usepackage{hyperref}
\hypersetup{
    colorlinks,
    citecolor=black,
    filecolor=black,
    linkcolor=red,
    urlcolor=black
}
\usepackage{theorem}
\usepackage{float}
\usepackage{ifthen}
\usepackage[OT1]{fontenc}

\newcommand{\R}{{\textsf{R}}}
\newcommand{\code}[1]{{\texttt{#1}}}
\newcommand{\term}[1]{{\emph{#1}}}
\newcommand{\Rpackage}[1]{\textsf{#1}}
\newcommand{\Rfunction}[1]{\texttt{#1}}
\newcommand{\Robject}[1]{\texttt{#1}}
\newcommand{\Rclass}[1]{{\textit{#1}}}
\newcommand{\Rmethod}[1]{{\textit{#1}}}
\newcommand{\Rfunarg}[1]{{\textit{#1}}}

\bibliographystyle{plainnat}
\title{cummeRbund: Visualization and Exploration of Cufflinks High-throughput Sequencing Data}

\author{Loyal A. Goff, Cole Trapnell}
\date{1 April, 2011}

\usepackage{Sweave}
\begin{document}


\maketitle

\tableofcontents

\section{Introduction}
	\Rpackage{cummeRbund} is a visualization package for Cufflinks high-throughput sequencing data. The base class, \Rclass{cuffSet} 
	is a 'pointer' to cufflinks data that are stored out-of-memory in a sqlite database.
	
\section{CummeRbund Classes}

\subsection{CuffSet Class}

\subsection{CuffData Class}

\subsection{CuffFeatureSet Class}

\subsection{CuffFeature Class}
	
\section{Reading cuffdiff output}
\begin{Schunk}
\begin{Sinput}
> cuff <- readCufflinks(dir = fileDir)
> cuff
\end{Sinput}
\begin{Soutput}
CuffSet instance with:
	 3 samples
	 35695 genes
	 102603 isoforms
	 54339 TSS
	 20584 CDS
\end{Soutput}
\end{Schunk}

\subsection{Reading additional annotation files}


\section{Global statistics}
Several plotting methods are available that allow for quality-control or global analysis of cufflinks data. For example, to assess the
distributions of FPKM scores across samples, you can use the \Rmethod{csDensity} plot (Figure 1).
\begin{Schunk}
\begin{Sinput}
> dens <- csDensity(cuff@genes)
\end{Sinput}
\end{Schunk}

\begin{figure}[ht]
\begin{center}
\includegraphics{cummeRbund-manual-global_plots_dens}

\end{center}
\caption{Density plot per sample of cufflinks output FPKM values.}
\end{figure}

Boxplots can be visualized using the \Rmethod{csBoxplot} method (Figure 2).
\begin{Schunk}
\begin{Sinput}
> b <- csBoxplot(cuff@genes)
\end{Sinput}
\end{Schunk}

\begin{figure}[ht]
\begin{center}
